linux-syscall/src/time.rs

```diff
@@ -1,32 +1,199 @@
+
+        Ok(0)
+    }
+
+    /// get time in seconds
+    #[cfg(target_arch = "x86_64")]
+    pub fn sys_time(&mut self, mut time: UserOutPtr<u64>) -> SysResult {
+        info!("time: time: {:?}", time);
+        let sec = TimeSpec::new().sec;
+        time.write(sec as u64)?;
+        Ok(sec)
+    }
+
+    /// get resource usage
+    /// currently only support ru_utime and ru_stime:
+    /// - `ru_utime`: user CPU time used
+    /// - `ru_stime`: system CPU time used
+    pub fn sys_getrusage(&mut self, who: usize, mut rusage: UserOutPtr<RUsage>) -> SysResult {
+        info!("getrusage: who: {}, rusage: {:?}", who, rusage);
+
+        let new_rusage = RUsage {
+            utime: TimeVal::new(),
+            stime: TimeVal::new(),
+        };
+        rusage.write(new_rusage)?;
+        Ok(0)
+    }
+
+    /// stores the current process times in the struct tms that buf points to
+    pub fn sys_times(&mut self, mut buf: UserOutPtr<Tms>) -> SysResult {
+        info!("times: buf: {:?}", buf);
+
+        let tv = TimeVal::new();
+
+        let tick = (tv.sec * 1_000_000 + tv.usec) / USEC_PER_TICK;
+
+        let new_buf = Tms {
+            tms_utime: 0,
+            tms_stime: 0,
+            tms_cutime: 0,
+            tms_cstime: 0,
+        };
+
+        buf.write(new_buf)?;
 
+        info!("tick: {:?}", tick);
+        Ok(tick as usize)
+    }
+}
+
+impl TimeVal {
+    /// create TimeVal
+    pub fn new() -> TimeVal {
+        TimeSpec::new().into()
+    }
+}
+
+impl TimeSpec {
+    /// create TimeSpec
+    pub fn new() -> TimeSpec {
         let time = timer_now();
-        let ts = TimeSpec {
+        TimeSpec {
             sec: time.as_secs() as usize,
             nsec: (time.as_nanos() % 1_000_000_000) as usize,
-        };
-        buf.write(ts)?;
+        }
+    }
 
-        Ok(0)
+    /// update TimeSpec for a file inode
+    /// TODO: more precise; update when write
+    pub fn update(inode: &Arc<dyn INode>) {
+        let now = TimeSpec::new().into();
+        if let Ok(mut metadata) = inode.metadata() {
+            metadata.atime = now;
+            metadata.mtime = now;
+            metadata.ctime = now;
+            // silently fail for device file
+            inode.set_metadata(&metadata).ok();
+        }
+    }
+}
+
+impl Into<Timespec> for TimeSpec {
+    fn into(self) -> Timespec {
+        Timespec {
+            sec: self.sec as i64,
+            nsec: self.nsec as i32,
+        }
+    }
+}
+
+impl Into<Duration> for TimeSpec {
+    fn into(self) -> Duration {
+        Duration::new(self.sec as u64, self.nsec as u32)
     }
 }
+
+impl Into<TimeVal> for TimeSpec {
+    fn into(self) -> TimeVal {
+        TimeVal {
+            sec: self.sec,
+            usec: self.nsec / 1_000 as usize,
+        }
+    }
+}
+
+impl Default for TimeVal {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Default for TimeSpec {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
Intuitively the default value of `TimeVal` or `TimeSpec` should be 0? Just like `Duration` does.